torch/distributed/algorithms/ddp_comm_hooks/mixed_precision_hooks.py

```diff
@@ -0,0 +1,82 @@
+import torch
+import torch.distributed as dist
+from torch.autograd import Variable
+
+from dataclasses import dataclass
+from typing import Any
+
+@dataclass
+class _AllreduceUpcastHookState:
+    """
+    State to manage DDP mixed precision in backward / gradient communication.
+    This contains a weakref to the DDP module for access to reducer and process
+    group, and a stream to run parameter and gradient upcasts.
+    """
+    ddp_weakref: Any
+    upcast_stream: torch.cuda.Stream
+    upcast_params_and_grads: bool
+    wait_for_stream_enqueued: bool = False
+
+def _reducer_allreduce_and_upcast_hook(
+    hook_state: _AllreduceUpcastHookState, bucket: dist.GradBucket
+) -> torch.futures.Future[torch.Tensor]:
+    """
+    Performs allreduce in the reduced precision given by DDP's mixed precision
+    reduce_dtype, and upcasts parameters and gradients to fp32 in preparation
+    to run the optimizer.
+    """
+    ddp_weakref = hook_state.ddp_weakref
+    reducer, process_group = ddp_weakref().reducer, ddp_weakref().process_group
+    gradient_is_bucket_view = ddp_weakref().gradient_as_bucket_view
+    # Cast bucket if different than param_dtype.
+    if (
+        ddp_weakref().mixed_precision.param_dtype != ddp_weakref().mixed_precision.reduce_dtype
+    ):
+        # Cast bucket tensor to reduce_dtype
+        bucket.set_buffer(bucket.buffer().to(ddp_weakref().mixed_precision.reduce_dtype))
+    fut = reducer._run_allreduce_hook(bucket)
+    ret_fut = torch.futures.Future()
+    stream = hook_state.upcast_stream
+    with torch.cuda.stream(stream):
+        fut.wait()
+        bucket.buffer().div_(process_group.size())
+        ret_fut.set_result(bucket.buffer())
```

Comment:
the allreduced bucket will be used to update grads in "finalize_bucket_dense()" later on, so the below upcast codes seem not working? the upcast codes possible should be done after "finalize_bucket_dense()"?

Reply:
hmm this is a good point, let me look into it

Reply:
The upcast code actually runs before `finalize_bucket_dense`: 

1. This code is kicked off in the comm hook callback, and run on a separate stream. This stream is waited on by the below queue_callback to the autograd engine. 
2. C++ finalize_backward also happens as an autograd engine final callback. However it runs _after_ the python callback, since we insert the python callback first, which is why the upcast code here works as expected. 
3. It is not great to rely on autograd engine callback running order. We can look into consolidating and inserting only a single callback always for DDP.